### ISSUE TITLE
3.5.0b1 - SSLEOFError

### ISSUE THREAD

**Issue by thezoggy** [sabnzbd/sabnzbd#1982]
sab 3.5.0b1 windows binary

was updating my servers with notes about how much blocks i have left (in preparation of BF deals) and afterwards saw a bunch of warnings.. unable to reproduce on demand (adding note/testing) but figured I'd note:

```
2021-11-20 10:41:44,975::DEBUG::[interface:130] Request GET /config/server/ from 192.168.0.2 [Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:94.0) Gecko/20100101 Firefox/94.0] {}
2021-11-20 10:41:45,477::INFO::[notifier:123] Sending notification: Warning - [20/Nov/2021:10:41:45] ENGINE socket.error 8
Traceback (most recent call last):
  File "cheroot\server.py", line 1287, in communicate
  File "cheroot\server.py", line 1077, in respond
  File "cheroot\wsgi.py", line 145, in respond
  File "cheroot\wsgi.py", line 231, in write
  File "cheroot\server.py", line 1133, in write
  File "cheroot\makefile.py", line 438, in write
  File "cheroot\makefile.py", line 36, in write
  File "cheroot\makefile.py", line 45, in _flush_unlocked
  File "socket.py", line 723, in write
  File "ssl.py", line 1205, in send
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2384)
 (type=warning, job_cat=None)
2021-11-20 10:41:45,477::WARNING::[_cplogging:213] [20/Nov/2021:10:41:45] ENGINE socket.error 8
Traceback (most recent call last):
  File "cheroot\server.py", line 1287, in communicate
  File "cheroot\server.py", line 1077, in respond
  File "cheroot\wsgi.py", line 145, in respond
  File "cheroot\wsgi.py", line 231, in write
  File "cheroot\server.py", line 1133, in write
  File "cheroot\makefile.py", line 438, in write
  File "cheroot\makefile.py", line 36, in write
  File "cheroot\makefile.py", line 45, in _flush_unlocked
  File "socket.py", line 723, in write
  File "ssl.py", line 1205, in send
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2384)

2021-11-20 10:41:45,477::WARNING::[_cplogging:213] [20/Nov/2021:10:41:45] ENGINE socket.error 8
Traceback (most recent call last):
  File "cheroot\server.py", line 1287, in communicate
  File "cheroot\server.py", line 1077, in respond
  File "cheroot\wsgi.py", line 145, in respond
  File "cheroot\wsgi.py", line 231, in write
  File "cheroot\server.py", line 1133, in write
  File "cheroot\makefile.py", line 438, in write
  File "cheroot\makefile.py", line 36, in write
  File "cheroot\makefile.py", line 45, in _flush_unlocked
  File "socket.py", line 723, in write
  File "ssl.py", line 1205, in send
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:2384)

2021-11-20 10:42:29,955::DEBUG::[interface:130] Request POST /config/server/saveServer from 192.168.0.2 [Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:94.0) Gecko/20100101 Firefox/94.0] {'apikey':<APIKEY>', 'output': 'json', 'server': 'news.usenet.farm', 'ajax': '1', 'news.usenet.farm': '1', 'enable': '1', 'displayname': 'Usenet.farm', 'host': 'news.usenet.farm', 'port': '563', 'ssl': '1', 'username': '<removed>', 'password': '**********', 'connections': '20', 'priority': '4', 'retention': '999', 'timeout': '120', 'ssl_verify': '2', 'ssl_ciphers': '', 'expire_date': '', 'quota': '', 'notes': xxx'}
```

I know this was a previously known python bug that should have been fixed in newer.. windows binary is packaged with py3.10
previous issue; https://github.com/sabnzbd/sabnzbd/issues/1577

**Comment by thezoggy** [comment]
looks like it may be: https://bugs.python.org/issue45487

**Comment by thezoggy** [comment]
config > special > allow_old_ssl_tls ( off )
which then is used:
https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/newswrapper.py#L287

looks like our context doesnt actually toss out the lower ssl anymore as we are using the legacy method

https://docs.python.org/3/library/ssl.html#context-creation
> Deprecated since version 3.7: The option is deprecated since OpenSSL 1.1.0, use the new SSLContext.minimum_version and SSLContext.maximum_version instead.

**Comment by puzzledsab** [comment]
On a side note, bulknews.eu is does not work when using Python 3.10. They are using TLSv1.2 (AES128-SHA). The warning is
`[Errno 10061] [SSL: SSLV3_ALERT_HANDSHAKE_FAILURE] sslv3 alert handshake failure (_ssl.c:997)`
Maybe you should add a warning to the release notes or something.

**Comment by Safihre** [comment]
Since we support only Python 3.7+ I switched now to the `SSLContext.minimum_version`.

The Bulknews problem seems bigger: all servers that only use TLSv1.2 stopped working because the ciphers are blocked. At least the ones I tried from @sanderjo's list: https://www.appelboor.com/newsservers/newsservers-with-SSL.html
I fear Python 3.10 is _too_ strict. We might have to manually create the context.

**Comment by Safihre** [comment]
Oh no.. Even when manually creating the context, the TLSv1.2 related ciphers are disabled.. Or at least I can't connect to any of the TLSv1.2 servers.

**Comment by puzzledsab** [comment]
It works by adding them using ctx.set_ciphers:
            self.nw.server.ssl_context.set_ciphers(
                "TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256:TLS_AES_128_GCM_SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES256-SHA:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:DHE-RSA-AES128-SHA:AES256-GCM-SHA384:AES128-GCM-SHA256:AES256-SHA256:AES128-SHA256:AES256-SHA:AES128-SHA"
            )
However, some are listed in 3.10 that aren't in 3.8, so I don't think we can use a fixed string. The ones that are supported would need to be added to the existing ones.

**Comment by Safihre** [comment]
We can also take the list-approach:
https://www.openssl.org/docs/manmaster/man1/openssl-ciphers.html
```python
self.nw.server.ssl_context.set_ciphers("MEDIUM:HIGH")
```
Or:
```python
self.nw.server.ssl_context.set_ciphers("HIGH:TLSv1.2")
```
Also works.

**Comment by thezoggy** [comment]
From that bug it sounded like py3.10 it wants to do  tls2.1+ by default..  which now somewhat invalidates the whole require modern tls option? So might need to rework to actually allow older by def since the default min now changed?

**Comment by Safihre** [comment]
Yes, that's why the "require modern TLS" option was replaced by "allow older TLS".

**Comment by thezoggy** [comment]
> Yes, that's why the "require modern TLS" option was replaced by "allow older TLS".

ok coo, sorry was viewing on phone and didn pick up the difference.

**Comment by sanderjo** [comment]
news.bulknews.eu gets a rating F from testssl.sh. So maybe gradually warn & later on not connect to such unsafe servers anymore ... ?

![image](https://user-images.githubusercontent.com/1273502/142772069-74e03ed8-7dd4-4af1-afb4-51f4f9712304.png)

![image](https://user-images.githubusercontent.com/1273502/142772107-5718b403-f5e0-44da-a8d0-1c9911f93be6.png)


```
 Rating (experimental) 

 Rating specs (not complete)  SSL Labs's 'SSL Server Rating Guide' (version 2009q from 2020-01-30)
 Specification documentation  https://github.com/ssllabs/research/wiki/SSL-Server-Rating-Guide
 Protocol Support (weighted)  0 (0)
 Key Exchange     (weighted)  0 (0)
 Cipher Strength  (weighted)  0 (0)
 Final Score                  0
 Overall Grade                F
 Grade cap reasons            Grade capped to F. Vulnerable to DROWN
                              Grade capped to F. SSLv2 is offered
                              Grade capped to C. Vulnerable to POODLE
                              Grade capped to B. TLS 1.1 offered
                              Grade capped to B. TLS 1.0 offered
                              Grade capped to B. SSLv3 is offered
                              Grade capped to B. RC4 ciphers offered
                              Grade capped to B. Forward Secrecy (FS) is not supported

 Done 2021-11-21 18:08:47 [ 174s] -->> 94.232.116.154:nntps (news.bulknews.eu) <<--

```

**Comment by sanderjo** [comment]
More fun with news.bulknews.eu:



sander@brixit:~/git/sabnzbd$ python3.8 -c "from sabnzbd.utils.happyeyeballs import happyeyeballs; print((happyeyeballs('news.bulknews.eu', port=563, 
use_ssl=True))) "
94.232.116.154

sander@brixit:~/git/sabnzbd$ python3.10 -c "from sabnzbd.utils.happyeyeballs import happyeyeballs; print((happyeyeballs('news.bulknews.eu', port=563,
 use_ssl=True))) "
None

sander@brixit:~/git/sabnzbd$ python3.8 -c "from sabnzbd.utils.happyeyeballs import happyeyeballs; print((happyeyeballs('newsreader.eweka.nl', port=56
3, use_ssl=True))) "
81.171.92.205

sander@brixit:~/git/sabnzbd$ python3.10 -c "from sabnzbd.utils.happyeyeballs import happyeyeballs; print((happyeyeballs('newsreader.eweka.nl', port=5
63, use_ssl=True))) "
81.171.92.219

**Comment by puzzledsab** [comment]
Eweka gets an F as well, at least when I test it on the web version of SSL Labs, so I assume that's common for usenet servers. I'm not sure, but I don't think those scores are relevant for connections using TLSv1.2 and AES128-SHA.

I didn't notice there were lists as well. Maybe `set_ciphers("MEDIUM:HIGH:TLSv1.2")` as default and `set_ciphers("ALL:SSLv3")` if `allow_old_ssl_tls` is true? That way users probably won't have to change their settings. I didn't see a way to enable SSLv2. I assume this can affect RSS feeds as well.

**Comment by jcfp** [comment]
> I'm not sure, but I don't think those scores are relevant for connections using TLSv1.2 and AES128-SHA.

These scores are indeed irrelevant and determined entirely by the insecure encryption algorithms offered server side. In reality, secure connections will negotiate the highest level shared between client and server regardless of the outdated and insecure garbage the latter offers on the lower end, as sab enforces a minimum tls level by default preventing a connection from being established if the server cannot meet that minimum.

**Comment by sanderjo** [comment]
> Eweka gets an F as well, at least when I test it on the web version of SSL Labs

How did you test that?

Because when I test with CLI tool testssl.sh, newsreader.eweka.nl gets a ... B. 

(Sidenote: The B is due to eweka still offering TLS 1.0 and TLS 1.1 (according to testssl, although I cannot reproduce with a CLI openssl command). I wonder if IRL those TLS-versions are still needed. Maybe Eweka administrators can check their logs if there are clients connected with those old protocols? Maybe a very old Synology or  Linux?
Because if eweka is willing to remove TLS 1.0 & 1.1, they could get a nice A ...)

```
./testssl.sh --ip one newsreader.eweka.nl:nntps
...

 Rating (experimental)

 Rating specs (not complete)  SSL Labs's 'SSL Server Rating Guide' (version 2009q from 2020-01-30)
 Specification documentation  https://github.com/ssllabs/research/wiki/SSL-Server-Rating-Guide
 Protocol Support (weighted)  95 (28)
 Key Exchange     (weighted)  90 (27)
 Cipher Strength  (weighted)  90 (36)
 Final Score                  91
 Overall Grade                B
 Grade cap reasons            Grade capped to B. TLS 1.1 offered
                              Grade capped to B. TLS 1.0 offered

 Done 2021-11-29 16:21:42 [ 306s] -->> 81.171.92.233:nntps (newsreader.eweka.nl) <<--
```


![image](https://user-images.githubusercontent.com/1273502/143895457-e1d20a48-7974-4469-905d-5b453d9ea298.png)

**Comment by puzzledsab** [comment]
> How did you test that?

ssllabs.com. I suspect that it found a web server on the same IP which it tested because when I tried it again now it wouldn't accept a custom port.

**Comment by Safihre** [comment]
Closed by 863c0e5
